### DISCUSSION TITLE
Does Ocelot support IIS virtual directories (load balancing)?

### DISCUSSION THREAD

**Discussion by NunoOlliveira** [ThreeMammals/Ocelot#1780] [Q&A] [answered]
Hello, and thank you for this awesome project.
I would like to know if there is any chance for supporting IIS virtual directories in a load balancing environment.

If you specify the virtual directory path in the downstream host, it will become an invalid path:
Example:
```json
"DownstreamHostAndPorts": [
        {
          "Host": "localhost/MyVirtualDirectory",
          "Port": 80
        },
        {
          "Host": "localhost/MyVirtualDirectory1",
          "Port": 80
        }
      ]
```
The URL will become `http://localhost/MyVirtualDirectory:80` instead of the `http://localhost:80/MyVirtualDirectory`.

Is this correct or am I missing something in the configuration file.

Best Regards.

**Comment by chkrishna2001** [comment]
Have you find any solution for this ?
I am having issue with this, port number is getting added after virtual directory and getting 

`A potentially dangerous Request.Path value was detected from the client (:).`

**Comment by NunoOlliveira** [comment]
Yes, we have changed 2 methods on the src\Ocelot\Request\Middleware\DownstreamRequest.cs file:

```csharp
        public HttpRequestMessage ToHttpRequestMessage()
        {
            Uri uri = new Uri(new UriBuilder() { Host = Host, Scheme = Scheme }.ToString());
            UriBuilder uriBuilder = new UriBuilder
            {
                Host = uri.Host,
                Path = (!string.IsNullOrEmpty(uri.AbsolutePath) && uri.AbsolutePath.Length > 3) ? $"{uri.AbsolutePath[1..^1]}{AbsolutePath}" : AbsolutePath,
                Query = RemoveLeadingQuestionMark(Query),
                Scheme = Scheme,
                Port = Port,
            };
            _request.RequestUri = uriBuilder.Uri;
            _request.Method = new HttpMethod(Method);
            return _request;
        }
```

and 

```csharp
        public string ToUri()
        {
            Uri uri = new Uri(new UriBuilder() { Host = Host, Scheme = Scheme }.ToString());
            UriBuilder uriBuilder = new UriBuilder
            {
                Host = uri.Host,
                Path = (!string.IsNullOrEmpty(uri.AbsolutePath) && uri.AbsolutePath.Length > 3) ? $"{uri.AbsolutePath[1..^1]}{AbsolutePath}" : AbsolutePath,
                Query = RemoveLeadingQuestionMark(Query),
                Scheme = Scheme,
                Port = Port,
            };
            return uriBuilder.Uri.AbsoluteUri;
        }
```

**Comment by raman-m** [comment, marked as answer]
Hi Nuno!
**Your routes definition is incorrect!** 
Ocelot does support virtual IIS directories, because virtual directory is a path after the host & port string.

I believe you have to define your routes like this:
```json
{
  "Routes": [
    {
      "UpstreamHttpMethod": [ "Get" ],
      "UpstreamPathTemplate": "/up-path",
      "DownstreamPathTemplate": "/MyVirtualDirectory",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        { "Host": "localhost", "Port": 80 }
      ]
    },
    {
      "UpstreamHttpMethod": [ "Get" ],
      "UpstreamPathTemplate": "/up-path1",
      "DownstreamPathTemplate": "/MyVirtualDirectory1",
      "DownstreamScheme": "http",
      "DownstreamHostAndPorts": [
        { "Host": "localhost", "Port": 80 }
      ]
    }
  ]
}
```
Pretty simple, right?

**Reply by raman-m** [reply]
Try to remove the port also.

**Comment by NunoOlliveira** [comment]
Thank you very much for your answer.

**Reply by raman-m** [reply]
You are welcome!
Stay with Ocelot! 🐯

**Comment by NunoOlliveira** [comment]
Hello, Unfortunately it doesn't work for the load balancing option.
In a normal configuration that is fine.

In your configuration example where should I put the "LoadBalancerOptions" for the two virtual directories?
Thank you and best regards.

**Reply by raman-m** [reply]
Please go away from virtual directories of IIS! 🙏 
Just go to Windows OS file `hosts` which locates in directory `c:\Windows\System32\drivers\etc\`!
Define temporary DNS records with hosts and given ports.
Go to IIS and convert your virtual dir apps to classic web app!
Redefine web app bindings (host & port) with values from `hosts` file.
Enjoy your development with Ocelot having real DNS records for "**DownstreamHostAndPorts**" properties in your routes. :wink:

This DevOps life hack will allow you to focus on Load Balancing without problems.

**Reply by NunoOlliveira** [reply]
_"Just go to Windows OS file hosts which locates in directory c:\Windows\System32\drivers\etc\!"_
OMG I just realized... great answer

**Comment by NunoOlliveira** [comment]
Thank you for your answer.
Unfortunately, that is not an option. 
The hack that I have made is working correctly, so I will keep as is.

Best Regards.

**Reply by raman-m** [reply]
Sorry?!... Why?
Don't you want to convert IIS virt dir apps to web ones? Really wonder!...
Do you administer these apps and IIS instance?

**Reply by NunoOlliveira** [reply]
Hello, 
It is not really virtual directories. 
In IIS you can have web applications inside other web applications running their own application pool (process).
![image](https://github.com/ThreeMammals/Ocelot/assets/60079921/11f8c562-8cc0-47d5-80d9-5b8b73f01c38)

The paths would be for example:
https://localhost/TestApplication
https://localhost/TestApplication/WebService1/
https://localhost/TestApplication/WebService2/
It give's you the possibility to have more organization on your application instead of having each service in distinct ports.

Best Regards

**Reply by raman-m** [reply]
Mr. Nuno,
I've [answered](https://github.com/ThreeMammals/Ocelot/discussions/1780#discussioncomment-8047793) you with possible fixes for this strange IIS usage to develop Ocelot app.
As a team, **we don't recommend to host Ocelot in IIS!** Such a deployment has a lot of [Gotchas](https://ocelot.readthedocs.io/en/latest/introduction/gotchas.html)!
Self-hosted deployment scenarios work better, like Console or Docker...

-----
> Does Ocelot support IIS virtual directories (load balancing)?

To start using Load Balancing you have to go away from IIS with must-have DNS setup! I have no idea how to use IIS and Load Balancing together. You need to read articles on Microsoft boards.

Sorry, Nuno, I cannot help you more with your DevOps IIS researches...
